### PR TITLE
Add bootstrap and deployment scripts

### PR DESCRIPTION
**Pull request by marrobi**
# PR for issue

## What is being addressed

Add scripts to bootstrap TF environment, build and push docker images, and deploy the TRE.

## How is this addressed

- Add TF bootstrap scripts and templates to create TF state storage and ACR
- Add scripts to build and push the docker images
- Add deployment scripts to core TF templates
- Remove GitHub packages integration
- Set linter to run locally as we have already checkout out the files

### CODE REVIEW

File: docs/quickstart.md [hunk 1]
```diff
@@ -45,29 +32,74 @@ az account list
...
+
+To build and push the docker images run:
+
+- From bash run `./build_and_publish_docker_images.sh.sh`
+
+This Terraform creates & configures the following:
+
+- Resource Group (also in bootstrap).
+- Storage Account for holding Terraform state (also in bootstrap).
+- Azure Container Registry.
```
Comment (deniscep) [inline docs/quickstart.md:86]: Looking at the code, this documentation doesn't match what the script does (build_and_publish.. doesn't create rg, etc.)?

File: templates/core/terraform/deploy.sh [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+if [[ -z $SUB_NAME ]]; then
+  echo -e "\n\e[31m»»» ⚠️ You are not logged in to Azure!"
+  exit
+fi
+export TENANT_ID=$(az account show --query tenantId -o tsv)
+
+echo -e "\e[34m»»» 🔨 \e[96mAzure details from logged on user \e[0m"
+echo -e "\e[34m»»»   • \e[96mSubscription: \e[33m$SUB_NAME\e[0m"
+echo -e "\e[34m»»»   • \e[96mTenant:       \e[33m$TENANT_ID\e[0m\n"
+
```
Comment (deniscep) [inline templates/core/terraform/deploy.sh:51]: same, copy-paste

File: templates/core/terraform/destroy.sh [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+if [[ -z $SUB_NAME ]]; then
+  echo -e "\n\e[31m»»» ⚠️ You are not logged in to Azure!"
+  exit
+fi
+export TENANT_ID=$(az account show --query tenantId -o tsv)
+
+echo -e "\e[34m»»» 🔨 \e[96mAzure details from logged on user \e[0m"
+echo -e "\e[34m»»»   • \e[96mSubscription: \e[33m$SUB_NAME\e[0m"
+echo -e "\e[34m»»»   • \e[96mTenant:       \e[33m$TENANT_ID\e[0m\n"
+
```
Comment (deniscep) [inline templates/core/terraform/destroy.sh:51]: copy-paste

**Review by deniscep — CHANGES_REQUESTED** [review]
Should we consider modularizing sh scripts, so we have less code? Following instructions the bootstrap.sh fails.

File: bootstrap/terraform/.env.sample [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# Used by backend state 
+TF_VAR_state_storage=__CHANGE_ME__
+TF_VAR_mgmt_res_group=__CHANGE_ME__
+TF_VAR_state_container=tfstate
+TF_VAR_location=__CHANGE_ME__
+TF_VAR_resource_name_prefix=__CHANGE_ME__
+`TF_VAR_image_tag=__CHANGE_ME__
```
[the remark this answers is not in the document]
Comment (marrobi) [reply bootstrap/terraform/.env.sample]: Weird, didn't show in my editor. Fixed.

File: bootstrap/terraform/bootstrap.sh [hunk 5]
```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (marrobi) [reply bootstrap/terraform/bootstrap.sh:1]: Fixed, are you ok to give it another go and I will look at modularising.

File: bootstrap/terraform/build_and_publish_docker_images.sh [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+export SUB_NAME=$(az account show --query name -o tsv)
+if [[ -z $SUB_NAME ]]; then
+  echo -e "\n\e[31m»»» ⚠️ You are not logged in to Azure!"
+  exit
+fi
+export TENANT_ID=$(az account show --query tenantId -o tsv)
+
+echo -e "\e[34m»»» 🔨 \e[96mAzure details from logged on user \e[0m"
+echo -e "\e[34m»»»   • \e[96mSubscription: \e[33m$SUB_NAME\e[0m"
+echo -e "\e[34m»»»   • \e[96mTenant:       \e[33m$TENANT_ID\e[0m\n"
```
[the remark this answers is not in the document]
Comment (marrobi) [reply bootstrap/terraform/build_and_publish_docker_images.sh]: Hmm, they are different, this one requires docker...

File: docs/quickstart.md (hunk 1 above)
Comment (marrobi) [reply docs/quickstart.md:86]: Sorry, more haste...

Comment (marrobi) [reply docs/quickstart.md:86]: Fixed

File: bootstrap/terraform/build_and_publish_docker_images.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (TessFerrandez) [reply bootstrap/terraform/build_and_publish_docker_images.sh]: Perhaps we can re-factor out the common code into one script in the interest of keeping things DRY

[the remark this answers is not in the document]
Comment (marrobi) [reply bootstrap/terraform/build_and_publish_docker_images.sh]: @TessFerrandez  I've just converted to draft, will refactor.

**Review by TessFerrandez — APPROVED** [review]
Approving with suggestion to modularize the deployment scripts

**Comment by marrobi** [comment]
@deniscep  @TessFerrandez  I've refactored the scripts, and added a makefile. If follow the doc, then can do everything by running `make xxxx`.

 I need to give it another run through, but let me know your thoughts.

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,49 @@
+.PHONY: bootstrap-init mgmt-deploy mgmt-destroy build_images push_images deploy-tre destroy-tre
+
+SHELL:=/bin/bash
+
+all: bootstrap mgmt-deploy build-images push-images tre-deploy
+
+bootstrap:
+	echo -e "\n\e[34m»»» 🧩 \e[96mBootstrap Terraform\e[0m..." \
+	&& . ./devops/scripts/check_dependancies.sh \
```
Comment (TessFerrandez) [inline Makefile:9]: nit: check_dependencies.sh

**Review by TessFerrandez — APPROVED** [review]
Looks great, just a small typo nit

**Comment by marrobi** [comment]
Tested end to end - if check out the repo, edit the env vars in './devops/terraform`, and run `make`, all stages needed to deploy from scratch will run.

**Review by christoferlof — CHANGES_REQUESTED** [review]
Overall smooth experience 👍, however I'm seeing an Terraform error on first deployment of a new instance.
```
╷
│ Error: Error: Public IP "pip-agw-<name>-<env>-<id>" (Resource Group "rg-<name>-<env>-<id>") was not found
│ 
│   with module.appgateway.data.azurerm_public_ip.appgwpip_data,
│   on appgateway/appgateway.tf line 82, in data "azurerm_public_ip" "appgwpip_data":
│   82: data "azurerm_public_ip" "appgwpip_data" {
│ 
╵
```
I created issue #94 to track and resolve separately.

File: docs/quickstart.md [hunk 8]
```diff
@@ -45,29 +32,64 @@ az account list
...
-cd templates/core/terraform
-terraform init
-```
+Copy [.env.sample](../devops/terraform/.env.sample) to `.env` in the  and set values for all variables:
 
-Create execution plan using your variables file created earlier. Specify a plan file name, we will use `azuretre-dev.tfplan`:
+- `TF_VAR_state_storage` - The name of the storage account to hold Terraform state.
+- `TF_VAR_mgmt_res_group` - The shared resource group for all hub resources, including the storage account.
+- `TF_VAR_state_container` - Name of the blob container to hold Terraform state (default: `tfstate`).
+- `TF_VAR_prefix` - A prefix added to all resources, pick your project name or other prefix to give the resources unique names.
```
[the remark this answers is not in the document]
Comment (marrobi) [reply docs/quickstart.md:46]: Such as the TRE project name, "hospitalx" or "christofer" for example. Can you thik of better wording?

File: docs/quickstart.md [hunk 9]
```diff
@@ -1,40 +1,27 @@
...
 The most straightforward way to get up and running is to deploy direct from the `microsoft/AzureTRE` repository. Production deployments should take advantage of your chosen DevOps CD tooling.
 
-### Clone the repository
+## Clone the repository
 Clone the repository to your local machine ( `git clone https://github.com/microsoft/AzureTRE.git` ) or you may choose to use our pre-configured dev container via GitHub Codespaces.
 
 ![Clone Options](../docs/assets/clone_options.png)
 
-### Pre-requisites
+## Pre-requisites
```
[the remark this answers is not in the document]
Comment (marrobi) [reply docs/quickstart.md:13]: Done

**Review by christoferlof — APPROVED** [review]

**Review by deniscep — APPROVED** [review]

**Pull request merged**